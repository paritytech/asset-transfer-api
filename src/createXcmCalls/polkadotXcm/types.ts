// Copyright 2023 Parity Technologies (UK) Ltd.

import type { XcmBaseArgs } from '../../types';

export type PolkadotXcmBaseArgs = XcmBaseArgs;
