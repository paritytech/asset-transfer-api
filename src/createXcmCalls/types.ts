// Copyright 2024 Parity Technologies (UK) Ltd.

export interface CreateXcmCallOpts {
	weightLimit?: { refTime?: string; proofSize?: string };
	paysWithFeeDest?: string;
	isLiquidTokenTransfer: boolean;
	isForeignAssetsTransfer: boolean;
	assetTransferType?: string;
	remoteReserveAssetTransferTypeLocation?: string;
	feesTransferType?: string;
	remoteReserveFeesTransferTypeLocation?: string;
	customXcmOnDest?: string;
}
