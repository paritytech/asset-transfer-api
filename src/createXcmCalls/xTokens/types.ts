import type { XcmBaseArgs } from '../../types';
import type { XcmPalletName } from '../util/establishXcmPallet';

export interface XTokensBaseArgs extends XcmBaseArgs {
	xcmPallet: XcmPalletName;
}
