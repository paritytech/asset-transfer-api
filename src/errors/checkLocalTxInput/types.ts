// Copyright 2023 Parity Technologies (UK) Ltd.

export enum LocalTxType {
	Assets = 'Assets',
	Balances = 'Balances',
	ForeignAssets = 'ForeignAssets',
	PoolAssets = 'PoolAssets',
	Tokens = 'Tokens',
}
