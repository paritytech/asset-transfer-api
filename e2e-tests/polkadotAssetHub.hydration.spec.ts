import { setupNetworks, testingPairs, withExpect } from '@acala-network/chopsticks-testing';
import { NetworkContext } from '@acala-network/chopsticks-utils';
import { AccountData } from '@polkadot/types/interfaces';
import { PalletAssetsAssetAccount } from '@polkadot/types/lookup';
import { Option } from '@polkadot/types-codec';
import { afterEach, beforeEach, expect, test } from 'vitest';

import { AssetTransferApi } from '../src/AssetTransferApi';

const { check } = withExpect(expect);

describe('Polkadot AssetHub <> Hydration', () => {
	let hydration: NetworkContext;
	let polkadotAssetHub: NetworkContext;

	const { alice } = testingPairs();
	const recipientAddress = '15McF4S5ZsoAJGzdXE3FwSFVjSPoz1Cd7Xj7VQZCb7HULcjx';
	const snowbridgeWETHLocation = {
		parents: 2,
		interior: {
			X2: [
				{ GlobalConsensus: { Ethereum: { chainId: 1 } } },
				{ AccountKey20: { network: null, key: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2' } },
			],
		},
	};

	beforeEach(async () => {
		const { hydration1, polkadotAssetHub1 } = await setupNetworks({
			hydration1: {
				endpoint: 'wss://hydration.ibp.network',
				db: './db.sqlite',
				port: 8006,
			},
			polkadotAssetHub1: {
				endpoint: 'wss://polkadot-asset-hub-rpc.polkadot.io',
				db: './db.sqlite',
				port: 8007,
			},
		});

		hydration = hydration1;
		polkadotAssetHub = polkadotAssetHub1;
	}, 500000);

	afterEach(async () => {
		await polkadotAssetHub.teardown();
		await hydration.teardown();
	}, 500000);

	describe('Local Transfers', () => {
		test('AssetHub Local DOT Transfer', async () => {
			await polkadotAssetHub.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // DOT
					],
				},
			});

			const recipientsInitialDOTBalance = await polkadotAssetHub.api.query.system.account(recipientAddress);

			await check(recipientsInitialDOTBalance).toMatchSnapshot('local assethub recipient initial dot balance');
			expect(recipientsInitialDOTBalance.data.free.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(polkadotAssetHub.api, 'asset-hub-polkadot', 4);
			const tx = await assetTransferApi.createTransferTransaction(
				'1000',
				recipientAddress,
				['dot'],
				['1000000000000'],
				{
					format: 'payload',
					sendersAddr: alice.address,
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await polkadotAssetHub.api.tx(extrinsic).signAndSend(alice);
			await polkadotAssetHub.dev.newBlock();

			const recipientsUpdatedDOTBalance = await polkadotAssetHub.api.query.system.account(recipientAddress);

			await check(recipientsUpdatedDOTBalance)
				.redact({ number: 2 })
				.toMatchSnapshot('local assethub recipient updated dot balance');
			expect(recipientsUpdatedDOTBalance.data.free.toNumber()).toBeGreaterThan(1000000);
		}, 200000);

		test('AssetHub Local Pool Asset Transfer', async () => {
			await polkadotAssetHub.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // DOT
						[[recipientAddress], { providers: 1, data: { free: 2 * 1e12 } }], // DOT
					],
				},
				PoolAssets: {
					Account: [[[30, alice.address], { balance: 75000000000000 }]],
				},
			});

			const aliceBalance = (await polkadotAssetHub.api.query.poolAssets.account(30, alice.address)).unwrapOrDefault();
			expect(aliceBalance.balance.toNumber()).toEqual(75000000000000);

			const recipientsInitialPoolAssetBalance = (
				await polkadotAssetHub.api.query.poolAssets.account(30, recipientAddress)
			).unwrapOrDefault();

			await check(recipientsInitialPoolAssetBalance).toMatchSnapshot(
				'local assethub recipients initial pool asset balance',
			);
			expect(recipientsInitialPoolAssetBalance.balance.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(polkadotAssetHub.api, 'asset-hub-polkadot', 4);
			const tx = await assetTransferApi.createTransferTransaction('1000', recipientAddress, ['30'], ['10'], {
				format: 'payload',
				sendersAddr: alice.address,
				transferLiquidToken: true,
			});

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await polkadotAssetHub.api.tx(extrinsic).signAndSend(alice);
			await polkadotAssetHub.dev.newBlock();

			const recipientsUpdatedPoolAssetBalance = (
				await polkadotAssetHub.api.query.poolAssets.account(30, recipientAddress)
			).unwrapOrDefault();

			await check(recipientsUpdatedPoolAssetBalance)
				.redact({ number: 2 })
				.toMatchSnapshot('local assethub recipients updated pool asset balance');
			expect(recipientsUpdatedPoolAssetBalance.balance.toNumber()).toBeGreaterThan(
				recipientsInitialPoolAssetBalance.balance.toNumber(),
			);
		}, 200000);

		test('Hydration Local HDX Transfer', async () => {
			await hydration.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // HDX
					],
				},
				Tokens: {
					Accounts: [
						[[alice.address, 0], { free: '50000000000000000000000000' }], // HDX
					],
				},
			});

			const recipientInitialHDXBalance = await hydration.api.query.tokens.accounts(recipientAddress, 0);
			await check(recipientInitialHDXBalance as AccountData).toMatchSnapshot(
				'local hydration recipient initial balance',
			);
			expect((recipientInitialHDXBalance as AccountData).free.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(hydration.api, 'hydradx', 4);
			const tx = await assetTransferApi.createTransferTransaction(
				'2034',
				recipientAddress,
				['0'], // HDX currency ID
				['1000000000000'],
				{
					format: 'payload',
					sendersAddr: alice.address,
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await hydration.api.tx(extrinsic).signAndSend(alice);
			await hydration.dev.newBlock();

			const recipientUpdatedHDXBalance = await hydration.api.query.tokens.accounts(recipientAddress, 0);

			await check(recipientUpdatedHDXBalance as AccountData)
				.redact({ number: 2 })
				.toMatchSnapshot('local hydration recipient updated balance');
			expect((recipientUpdatedHDXBalance as AccountData).free.toNumber()).toBeGreaterThan(
				(recipientInitialHDXBalance as AccountData).free.toNumber(),
			);
		}, 200000);
	});

	describe('XCM V3', () => {
		const xcmVersion = 3;

		test('Transfer DOT From Hydration to AssetHub', async () => {
			await hydration.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // HDX
					],
				},
				Tokens: {
					Accounts: [
						[[alice.address, 5], { free: 1000 * 1e12 }], // DOT
					],
				},
			});

			const recipientsInitialDOTBalance = await polkadotAssetHub.api.query.system.account(recipientAddress);

			await check(recipientsInitialDOTBalance).toMatchSnapshot('asset hub recipients initial dot balance');
			expect(recipientsInitialDOTBalance.data.free.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(hydration.api, 'hydradx', xcmVersion);
			const tx = await assetTransferApi.createTransferTransaction(
				'1000',
				recipientAddress,
				['DOT'],
				['1000000000000'],
				{
					format: 'payload',
					xcmVersion,
					sendersAddr: alice.address,
				},
			);
			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await hydration.api.tx(extrinsic).signAndSend(alice);
			await hydration.dev.newBlock();
			await polkadotAssetHub.dev.newBlock();

			const recipientsUpdatedDOTBalance = await polkadotAssetHub.api.query.system.account(recipientAddress);

			await check(recipientsUpdatedDOTBalance)
				.redact({ number: 2 })
				.toMatchSnapshot('asset hub recipients updated dot balance');
			expect(recipientsUpdatedDOTBalance.data.free.toNumber()).toBeGreaterThan(
				recipientsInitialDOTBalance.data.free.toNumber(),
			);
		}, 200000);

		test('Transfer DOT From AssetHub to Hydration', async () => {
			await polkadotAssetHub.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // DOT
					],
				},
			});

			const recipientsInitialDOTBalance = await hydration.api.query.tokens.accounts(recipientAddress, 5);

			await check(recipientsInitialDOTBalance as AccountData).toMatchSnapshot(
				'hydration recipients initial dot balance',
			);
			expect((recipientsInitialDOTBalance as AccountData).free.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(polkadotAssetHub.api, 'asset-hub-polkadot', xcmVersion);
			const tx = await assetTransferApi.createTransferTransaction(
				'2034',
				recipientAddress,
				['DOT'],
				['1000000000000'],
				{
					format: 'payload',
					xcmVersion,
					sendersAddr: alice.address,
				},
			);
			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await polkadotAssetHub.api.tx(extrinsic).signAndSend(alice);
			await polkadotAssetHub.dev.newBlock();
			await hydration.dev.newBlock();

			const recipientsUpdatedDOTBalance = await hydration.api.query.tokens.accounts(recipientAddress, 5);

			await check(recipientsUpdatedDOTBalance as AccountData)
				.redact({ number: 2 })
				.toMatchSnapshot('hydration recipients updated dot balance');
			expect((recipientsUpdatedDOTBalance as AccountData).free.toNumber()).toBeGreaterThan(
				(recipientsInitialDOTBalance as AccountData).free.toNumber(),
			);
		}, 200000);

		test('Transfer SnowBridge WETH From Hydration To AssetHub', async () => {
			await hydration.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // HDX
					],
				},
				Tokens: {
					Accounts: [
						[[alice.address, 0], { free: '50000000000000000000000000' }], // HDX
						[[alice.address, 1000189], { free: '50000000000000000000000000' }], // Snowbridge WETH
					],
				},
			});

			const recipientsInitialWETHBalance = (
				(await polkadotAssetHub.api.query.foreignAssets.account(
					snowbridgeWETHLocation,
					recipientAddress,
				)) as Option<PalletAssetsAssetAccount>
			).unwrapOrDefault();

			await check(recipientsInitialWETHBalance).toMatchSnapshot('asset hub recipients initial snowbridge weth balance');
			expect(recipientsInitialWETHBalance.balance.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(hydration.api, 'hydradx', xcmVersion, {
				registryType: 'NPM',
				injectedRegistry: {
					polkadot: {
						2034: {
							tokens: [],
							assetsInfo: {},
							foreignAssetsInfo: {},
							poolPairsInfo: {},
							specName: 'hydradx',
							xcAssetsData: [
								{
									paraID: 0,
									symbol: 'WETH.snow',
									decimals: 18,
									xcmV1MultiLocation:
										'{"v1":{"parents":2,"interior":{"x2":[{"globalConsensus":{"ethereum":{"chainId":1}}},{"accountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}}',
									asset: '1000189',
									assetHubReserveLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
								},
							],
						},
					},
				},
			});
			const tx = await assetTransferApi.createTransferTransaction(
				'1000',
				recipientAddress,
				['WETH.snow'],
				['25000000000000'],
				{
					format: 'payload',
					xcmVersion,
					sendersAddr: '7NPoMQbiA6trJKkjB35uk96MeJD4PGWkLQLH7k7hXEkZpiba',
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await hydration.api.tx(extrinsic).signAndSend(alice);
			await hydration.dev.newBlock();

			await polkadotAssetHub.dev.newBlock();

			const recipientsUpdatedWETHBalance = (
				(await polkadotAssetHub.api.query.foreignAssets.account(
					snowbridgeWETHLocation,
					recipientAddress,
				)) as Option<PalletAssetsAssetAccount>
			).unwrapOrDefault();

			await check(recipientsUpdatedWETHBalance)
				.redact({ number: 1 })
				.toMatchSnapshot('asset hub recipients updated snowbridge weth balance');
			expect(recipientsUpdatedWETHBalance.balance.toNumber()).toBeGreaterThan(
				recipientsInitialWETHBalance.balance.toNumber(),
			);
		}, 200000);

		test('Transfer SnowBridge WETH From AssetHub To Hydration', async () => {
			await polkadotAssetHub.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // DOT
					],
				},
				ForeignAssets: {
					Account: [[[snowbridgeWETHLocation, alice.address], { balance: 75000000000000 }]],
				},
			});

			const recipientsInitialHydrationWETHBalance = await hydration.api.query.tokens.accounts(
				recipientAddress,
				1000189,
			);

			await check(recipientsInitialHydrationWETHBalance as AccountData).toMatchSnapshot(
				'hydration recipients initial snowbridge weth balance',
			);
			expect((recipientsInitialHydrationWETHBalance as AccountData).free.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(polkadotAssetHub.api, 'asset-hub-polkadot', xcmVersion);
			const tx = await assetTransferApi.createTransferTransaction(
				'2034',
				recipientAddress,
				[
					`{"parents":"2","interior":{"X2":[{"GlobalConsensus":{"Ethereum":{"chainId":"1"}}},{"AccountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}`,
				],
				['25000000000000'],
				{
					format: 'payload',
					xcmVersion,
					sendersAddr: alice.address,
					paysWithFeeDest:
						'{"parents":"2","interior":{"X2":[{"GlobalConsensus":{"Ethereum":{"chainId":"1"}}},{"AccountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}',
					assetTransferType: 'LocalReserve',
					remoteReserveAssetTransferTypeLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
					feesTransferType: 'LocalReserve',
					remoteReserveFeesTransferTypeLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await polkadotAssetHub.api.tx(extrinsic).signAndSend(alice);
			await polkadotAssetHub.dev.newBlock();
			await hydration.dev.newBlock();

			const recipientsUpdatedWETHBalance = await hydration.api.query.tokens.accounts(recipientAddress, 1000189);

			await check(recipientsUpdatedWETHBalance as AccountData)
				.redact({ number: 2 })
				.toMatchSnapshot('hydration recipients updated snowbridge weth balance');
			expect((recipientsUpdatedWETHBalance as AccountData).free.toNumber()).toBeGreaterThan(
				(recipientsInitialHydrationWETHBalance as AccountData).free.toNumber(),
			);
		}, 200000);
	});

	describe('XCM V4', () => {
		const xcmVersion = 4;

		test('Transfer DOT From Hydration to AssetHub', async () => {
			await hydration.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // HDX
					],
				},
				Tokens: {
					Accounts: [
						[[alice.address, 5], { free: 1000 * 1e12 }], // DOT
					],
				},
			});

			const recipientsInitialDOTBalance = await polkadotAssetHub.api.query.system.account(recipientAddress);

			await check(recipientsInitialDOTBalance).toMatchSnapshot('asset hub recipients initial dot balance');
			expect(recipientsInitialDOTBalance.data.free.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(hydration.api, 'hydradx', xcmVersion);
			const tx = await assetTransferApi.createTransferTransaction(
				'1000',
				recipientAddress,
				['DOT'],
				['1000000000000'],
				{
					format: 'payload',
					xcmVersion,
					sendersAddr: alice.address,
				},
			);
			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await hydration.api.tx(extrinsic).signAndSend(alice);
			await hydration.dev.newBlock();
			await polkadotAssetHub.dev.newBlock();

			const recipientsUpdatedDOTBalance = await polkadotAssetHub.api.query.system.account(recipientAddress);

			await check(recipientsUpdatedDOTBalance)
				.redact({ number: 2 })
				.toMatchSnapshot('asset hub recipients updated dot balance');
			expect(recipientsUpdatedDOTBalance.data.free.toNumber()).toBeGreaterThan(
				recipientsInitialDOTBalance.data.free.toNumber(),
			);
		}, 200000);

		test('Transfer DOT From AssetHub to Hydration', async () => {
			await polkadotAssetHub.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // DOT
					],
				},
			});

			const recipientsInitialDOTBalance = await hydration.api.query.tokens.accounts(recipientAddress, 5);

			await check(recipientsInitialDOTBalance as AccountData).toMatchSnapshot(
				'hydration recipients initial dot balance',
			);
			expect((recipientsInitialDOTBalance as AccountData).free.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(polkadotAssetHub.api, 'asset-hub-polkadot', xcmVersion);
			const tx = await assetTransferApi.createTransferTransaction(
				'2034',
				recipientAddress,
				['DOT'],
				['1000000000000'],
				{
					format: 'payload',
					xcmVersion,
					sendersAddr: alice.address,
				},
			);
			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await polkadotAssetHub.api.tx(extrinsic).signAndSend(alice);
			await polkadotAssetHub.dev.newBlock();
			await hydration.dev.newBlock();

			const recipientsUpdatedDOTBalance = await hydration.api.query.tokens.accounts(recipientAddress, 5);

			await check(recipientsUpdatedDOTBalance as AccountData)
				.redact({ number: 2 })
				.toMatchSnapshot('hydration recipients updated dot balance');
			expect((recipientsUpdatedDOTBalance as AccountData).free.toNumber()).toBeGreaterThan(
				(recipientsInitialDOTBalance as AccountData).free.toNumber(),
			);
		}, 200000);

		test('Transfer SnowBridge WETH From Hydration To AssetHub', async () => {
			await hydration.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // HDX
					],
				},
				Tokens: {
					Accounts: [
						[[alice.address, 0], { free: '50000000000000000000000000' }], // HDX
						[[alice.address, 1000189], { free: '50000000000000000000000000' }], // Snowbridge WETH
					],
				},
			});

			const recipientsInitialWETHBalance = (
				(await polkadotAssetHub.api.query.foreignAssets.account(
					snowbridgeWETHLocation,
					recipientAddress,
				)) as Option<PalletAssetsAssetAccount>
			).unwrapOrDefault();

			await check(recipientsInitialWETHBalance).toMatchSnapshot('asset hub recipients initial snowbridge weth balance');
			expect(recipientsInitialWETHBalance.balance.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(hydration.api, 'hydradx', xcmVersion, {
				registryType: 'NPM',
				injectedRegistry: {
					polkadot: {
						2034: {
							tokens: [],
							assetsInfo: {},
							foreignAssetsInfo: {},
							poolPairsInfo: {},
							specName: 'hydradx',
							xcAssetsData: [
								{
									paraID: 0,
									symbol: 'WETH.snow',
									decimals: 18,
									xcmV1MultiLocation:
										'{"v1":{"parents":2,"interior":{"x2":[{"globalConsensus":{"ethereum":{"chainId":1}}},{"accountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}}',
									asset: '1000189',
									assetHubReserveLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
								},
							],
						},
					},
				},
			});
			const tx = await assetTransferApi.createTransferTransaction(
				'1000',
				recipientAddress,
				['WETH.snow'],
				['25000000000000'],
				{
					format: 'payload',
					xcmVersion,
					sendersAddr: '7NPoMQbiA6trJKkjB35uk96MeJD4PGWkLQLH7k7hXEkZpiba',
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await hydration.api.tx(extrinsic).signAndSend(alice);
			await hydration.dev.newBlock();

			await polkadotAssetHub.dev.newBlock();

			const recipientsUpdatedWETHBalance = (
				(await polkadotAssetHub.api.query.foreignAssets.account(
					snowbridgeWETHLocation,
					recipientAddress,
				)) as Option<PalletAssetsAssetAccount>
			).unwrapOrDefault();

			await check(recipientsUpdatedWETHBalance)
				.redact({ number: 2 })
				.toMatchSnapshot('asset hub recipients updated snowbridge weth balance');
			expect(recipientsUpdatedWETHBalance.balance.toNumber()).toBeGreaterThan(
				recipientsInitialWETHBalance.balance.toNumber(),
			);
		}, 200000);

		test('Transfer SnowBridge WETH From AssetHub To Hydration', async () => {
			await polkadotAssetHub.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // DOT
					],
				},
				ForeignAssets: {
					Account: [[[snowbridgeWETHLocation, alice.address], { balance: 75000000000000 }]],
				},
			});

			const recipientsInitialHydrationWETHBalance = await hydration.api.query.tokens.accounts(
				recipientAddress,
				1000189,
			);

			await check(recipientsInitialHydrationWETHBalance as AccountData).toMatchSnapshot(
				'hydration recipients initial snowbridge weth balance',
			);
			expect((recipientsInitialHydrationWETHBalance as AccountData).free.toNumber()).toEqual(0);

			const assetTransferApi = new AssetTransferApi(polkadotAssetHub.api, 'asset-hub-polkadot', xcmVersion);
			const tx = await assetTransferApi.createTransferTransaction(
				'2034',
				recipientAddress,
				[
					`{"parents":"2","interior":{"X2":[{"GlobalConsensus":{"Ethereum":{"chainId":"1"}}},{"AccountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}`,
				],
				['25000000000000'],
				{
					format: 'payload',
					xcmVersion,
					sendersAddr: alice.address,
					paysWithFeeDest:
						'{"parents":"2","interior":{"X2":[{"GlobalConsensus":{"Ethereum":{"chainId":"1"}}},{"AccountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}',
					assetTransferType: 'LocalReserve',
					remoteReserveAssetTransferTypeLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
					feesTransferType: 'LocalReserve',
					remoteReserveFeesTransferTypeLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await polkadotAssetHub.api.tx(extrinsic).signAndSend(alice);
			await polkadotAssetHub.dev.newBlock();
			await hydration.dev.newBlock();

			const recipientsUpdatedWETHBalance = await hydration.api.query.tokens.accounts(recipientAddress, 1000189);

			await check(recipientsUpdatedWETHBalance as AccountData)
				.redact({ number: 2 })
				.toMatchSnapshot('hydration recipients updated snowbridge weth balance');
			expect((recipientsUpdatedWETHBalance as AccountData).free.toNumber()).toBeGreaterThan(
				(recipientsInitialHydrationWETHBalance as AccountData).free.toNumber(),
			);
		}, 200000);
	});
});
