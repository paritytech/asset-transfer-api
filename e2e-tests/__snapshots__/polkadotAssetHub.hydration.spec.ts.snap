// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Polkadot AssetHub <> Hydration > Local Transfers > AssetHub Local DOT Transfer > local assethub recipient initial dot balance 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": 0,
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 0,
  "sufficients": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > Local Transfers > AssetHub Local DOT Transfer > local assethub recipient updated dot balance 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": 1000000000000,
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > Local Transfers > AssetHub Local Pool Asset Transfer > local assethub recipients initial pool asset balance 1`] = `
{
  "balance": 0,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`Polkadot AssetHub <> Hydration > Local Transfers > AssetHub Local Pool Asset Transfer > local assethub recipients updated pool asset balance 1`] = `
{
  "balance": 10,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`Polkadot AssetHub <> Hydration > Local Transfers > Hydration Local HDX Transfer > local hydration recipient initial balance 1`] = `
{
  "free": 0,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > Local Transfers > Hydration Local HDX Transfer > local hydration recipient updated balance 1`] = `
{
  "free": 1000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V3 > Transfer DOT From AssetHub to Hydration > hydration recipients initial dot balance 1`] = `
{
  "free": 0,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V3 > Transfer DOT From AssetHub to Hydration > hydration recipients updated dot balance 1`] = `
{
  "free": "(rounded 1000000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V3 > Transfer DOT From Hydration to AssetHub > asset hub recipients initial dot balance 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": 0,
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 0,
  "sufficients": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V3 > Transfer DOT From Hydration to AssetHub > asset hub recipients updated dot balance 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": "(rounded 1000000000000)",
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V3 > Transfer SnowBridge WETH From AssetHub To Hydration > hydration recipients initial snowbridge weth balance 1`] = `
{
  "free": 0,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V3 > Transfer SnowBridge WETH From AssetHub To Hydration > hydration recipients updated snowbridge weth balance 1`] = `
{
  "free": "(rounded 24500000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V3 > Transfer SnowBridge WETH From Hydration To AssetHub > asset hub recipients initial snowbridge weth balance 1`] = `
{
  "balance": 0,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V3 > Transfer SnowBridge WETH From Hydration To AssetHub > asset hub recipients updated snowbridge weth balance 1`] = `
{
  "balance": "(rounded 17300000000000)",
  "extra": null,
  "reason": {
    "sufficient": null,
  },
  "status": "Liquid",
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V4 > Transfer DOT From AssetHub to Hydration > hydration recipients initial dot balance 1`] = `
{
  "free": 0,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V4 > Transfer DOT From AssetHub to Hydration > hydration recipients updated dot balance 1`] = `
{
  "free": "(rounded 1000000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V4 > Transfer DOT From Hydration to AssetHub > asset hub recipients initial dot balance 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": 0,
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 0,
  "sufficients": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V4 > Transfer DOT From Hydration to AssetHub > asset hub recipients updated dot balance 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": "(rounded 1000000000000)",
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V4 > Transfer SnowBridge WETH From AssetHub To Hydration > hydration recipients initial snowbridge weth balance 1`] = `
{
  "free": 0,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V4 > Transfer SnowBridge WETH From AssetHub To Hydration > hydration recipients updated snowbridge weth balance 1`] = `
{
  "free": "(rounded 24500000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V4 > Transfer SnowBridge WETH From Hydration To AssetHub > asset hub recipients initial snowbridge weth balance 1`] = `
{
  "balance": 0,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`Polkadot AssetHub <> Hydration > XCM V4 > Transfer SnowBridge WETH From Hydration To AssetHub > asset hub recipients updated snowbridge weth balance 1`] = `
{
  "balance": "(rounded 17300000000000)",
  "extra": null,
  "reason": {
    "sufficient": null,
  },
  "status": "Liquid",
}
`;
